csrc/vmm/vmm.cu

```diff
@@ -0,0 +1,242 @@
+
+// get CUdeviceptr dptr
+CUdeviceptr CacheDevicePtr::get_dptr() { return dptr; }
+
+// get void * type pointer
+void* CacheDevicePtr::get_void_ptr() { return reinterpret_cast<void*>(dptr); }
+
+/*
+** CacheAllocator functions implementation
+*/
+
+CacheAllocator::CacheAllocator() {
+  // get current device gpu id
+  int currentDevice;
+  auto cudaStatus = cudaGetDevice(&currentDevice);
+  TORCH_CHECK(cudaStatus == cudaSuccess, "cudaGetDevice failed!");
+
+  // set memory allocation property struct CUmemAllocationProp,
+  // which is used to control the specific behavior of memory allocation
+  prop = {};
+  prop.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+  prop.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+  prop.location.id = currentDevice;
+
+  // set memory access descriptor struct CUmemAccessDesc,
+  // which is used to control the access permission of memory
+  accessDescr = {};
+  accessDescr.location.id = prop.location.id;
+  accessDescr.location.type = prop.location.type;
+  accessDescr.flags = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+
+  // get the granularity of device memory allocation
+  // getGranurality();
+
+  // set the page size of memory allocation, default is equal to granurality
+  // pageSize = granurality;
+}
+
+int64_t CacheAllocator::getGranurality() {
+  cuMemGetAllocationGranularity(&granurality, &prop,
+                                CU_MEM_ALLOC_GRANULARITY_MINIMUM);
+  printf("Granularity: %ld Bytes\n", granurality);
+  return granurality;
+}
+
+void CacheAllocator::setPageSize(int64_t num) { pageSize = num * granurality; }
+
+// reserve function, reserve virtual address space
+int64_t CacheAllocator::reserveCachePtr(
+    const c10::intrusive_ptr<CacheDevicePtr>& ptr, int64_t pageNum) {
+  if (pageNum == 0) {
+    return CUDA_SUCCESS;
+  }
+  size_t size = pageNum * pageSize;
+  auto status = cuMemAddressReserve(&(ptr->dptr), size, 0, 0, 0);
+
+  if (status != CUDA_SUCCESS) {
+    printf("cuMemAddressReserve failed! error-code: %d\n", status);
+  } else {
+    ptr->reservedPageNum += pageNum;
+  }
+
+  return status;
+}
+
+// alloc function, allocate physical memory, map to the reserved virtual address
+// space of dptr, and set access permission
+int64_t CacheAllocator::allocCachePtr(
+    const c10::intrusive_ptr<CacheDevicePtr>& ptr, int64_t pageNum,
+    int64_t offset) {
+  if (pageNum == 0) {
+    return CUDA_SUCCESS;
+  }
+  // size = ((size - 1) / pageSize + 1) * pageSize;
+  size_t size = pageNum * pageSize;
+  auto start_dptr = ptr->dptr + offset;
+
+  CUresult status = CUDA_SUCCESS;
+  CUmemGenericAllocationHandle allocationHandle;
+  if ((status = cuMemCreate(&allocationHandle, size, &prop, 0)) ==
+      CUDA_SUCCESS) {
+    if ((status = cuMemMap(start_dptr, size, 0, allocationHandle, 0)) ==
+        CUDA_SUCCESS) {
+      if ((status = cuMemSetAccess(start_dptr, size, &accessDescr, 1)) ==
+          CUDA_SUCCESS) {
+        // ptr.handles.push_back(allocationHandle);  // handles is unused now
```

Comment:
Can you explain when we would use it?

Reply:
handle can be shared using the cuMemExportToShareableHandle api, and may be able to be used in a prefix cache in future work.

Also, very sorry for some of the cruft issues below, I'll check my code again once more

Reply:
No worries, appreciate your cleanup so we can better consider landing this.